Pkcs11 labeled token

mwasilew opened the pull request.

mwasilew committed:
[fio toup] crypto: make aktualizr find pkcs11 token by label

This prevents aktualizr from using 1st initialized pkcs11 token and
forces it to use the one with given label. Label is defined in [p11]
section of the config.

Signed-off-by: Milosz Wasilewski <milosz.wasilewski@foundries.io>

mwasilew committed:
[fio toup] crypto: update tests

Signed-off-by: Milosz Wasilewski <milosz.wasilewski@foundries.io>

ricardosalveti reviewed (approved):
LGTM (pkcs11 logic).

mike-sul commented:
Just two organizational comments:
1. Please recreate the PR against `2022.10+fio` branch (appology for not advising it in the first place).
2. Add the prefix `[fio toup]` to each commit message.

mwasilew commented:
@mike-sul I rebased on top of 2022.10+fio and modified the commit messages. I also changed PR base in github, but I'm not sure if it works properly. If needed I'll close this PR and start a new one against 2022.10+fio

mike-sul commented on src/libaktualizr/crypto/p11engine.cc:
```diff
@@ -145,7 +145,25 @@ boost::filesystem::path P11Engine::findPkcsLibrary() {
+  PKCS11_SLOT* slot = NULL;
+  PKCS11_SLOT* iterslot = NULL;
+  PKCS11_TOKEN* tok;
+  nslot = wslots_.get_nslots();
```
Just minor: `nslot` better define as `const` - `const auto nslot{wslots_.get_nslots()};`

mike-sul commented on src/libaktualizr/crypto/p11engine.cc:
```diff
@@ -145,7 +145,25 @@ boost::filesystem::path P11Engine::findPkcsLibrary() {
+    LOG_ERROR << "Token label missing. Unable to select slot.";
+    return nullptr;
+  }
+  for (i = 0, iterslot = wslots_.get_slots(); i < nslot; i++, iterslot++) {
```
Just minor. We may wanna define  `iterslot` before the loop `auto iterslot{wslots_.get_slots()}` and then define `i` within the loop `for (unsigned int i; i < nslot...`.

mike-sul commented on src/libaktualizr/crypto/p11engine.cc:
```diff
@@ -145,7 +145,25 @@ boost::filesystem::path P11Engine::findPkcsLibrary() {
+  }
+  for (i = 0, iterslot = wslots_.get_slots(); i < nslot; i++, iterslot++) {
+    if ((tok = iterslot->token) != NULL) {
+      if (label_.compare(tok->label) == 0) {
```
We can just do `label_ == tok->label`.

mike-sul commented on src/libaktualizr/crypto/p11engine.cc:
```diff
@@ -145,7 +145,25 @@ boost::filesystem::path P11Engine::findPkcsLibrary() {
+    return nullptr;
+  }
+  for (i = 0, iterslot = wslots_.get_slots(); i < nslot; i++, iterslot++) {
+    if ((tok = iterslot->token) != NULL) {
```
We can double check whether `iterslot is not NULL before the `if`.

mike-sul commented on src/libaktualizr/crypto/p11engine.cc:
```diff
@@ -145,7 +145,25 @@ boost::filesystem::path P11Engine::findPkcsLibrary() {
 PKCS11_SLOT* P11Engine::findTokenSlot() const {
-  PKCS11_SLOT* slot = PKCS11_find_token(ctx_.get(), wslots_.get_slots(), wslots_.get_nslots());
+  unsigned int i, nslot;
+  PKCS11_SLOT* slot = NULL;
```
Please, use `nullptr` instead of NULL for sake of consistency with the rest of the code.
Also, use the uniform initializer `PKCS11_SLOT* slot{nullptr}`

mike-sul commented on src/libaktualizr/crypto/p11engine.cc:
```diff
@@ -145,7 +145,25 @@ boost::filesystem::path P11Engine::findPkcsLibrary() {
+    }
+  }
   if ((slot == nullptr) || (slot->token == nullptr)) {
     LOG_ERROR << "Couldn't find a token";
```
```suggestion
    LOG_ERROR << "Couldn't find a token with the label " << label_;
```

mike-sul commented on src/libaktualizr/crypto/p11engine.cc:
```diff
@@ -145,7 +145,25 @@ boost::filesystem::path P11Engine::findPkcsLibrary() {
+  PKCS11_TOKEN* tok;
+  nslot = wslots_.get_nslots();
+
+  if (label_.empty()) {
```
I think we need to fallback to the previous search approach `PKCS11_SLOT* slot = PKCS11_find_token(ctx_.get(), wslots_.get_slots(), wslots_.get_nslots())` if the label is empty for backward compatibility.  

Many companies use libaktualizr and obviously they do not have this new config param (`label`) set in sota.toml so returning `nullptr` if the `label` is not specified will break their libaktualizr-based client.

mwasilew replied:
I guess you're right, but IMHO returning any initialized token is a bug. It might work if aktualizr is the only thing using pkcs11, but it's an incorrect approach. One mitigation might be attempt to access the token with pin. If this works there is a smaller chance of accessing wrong token.

mike-sul replied:
I agree that the current approach is not quite correct, however it has became a "feature bug/issue" already :). So we have to make it backward compatible.

mike-sul replied:
We can upstream the version with the backward compatibility and then add another commit with prefix [fio extra] that removes it and yields an error if no label.

mike-sul replied:
But, it most like will break the existing devices that are not configured with any label and use HSM.

ricardosalveti replied:
At least I believe all devices registered via lmp-device-register will have a valid 'aktualizr' token label, then I'm not so sure who else might be using without lmp-device-register (but it is indeed good to not break previous users).

mwasilew replied:
I just fixed it to be backward compatible - working with the same bug as before.

mike-sul replied:
>At least I believe all devices registered via lmp-device-register will have a valid 'aktualizr' token label,

It's about the new label param of the sota config, not the pkcs11 slot label. The currently deployed devices do not have it.

mike-sul commented on include/libaktualizr/config.h:
```diff
@@ -28,6 +28,7 @@ struct PKCS11_slot_st;
 struct P11Config {
   boost::filesystem::path module;
   std::string pass;
+  std::string label;
```
I think adding of a new param requires its serialization/deserialization so it can be actually read from `sota.toml`, something like this https://github.com/foundriesio/aktualizr/blob/1daf2d5a6bc8b3fdef926922cf17bd0685f187bc/src/libaktualizr/config/base_config.cc#L48 and https://github.com/foundriesio/aktualizr/blob/1daf2d5a6bc8b3fdef926922cf17bd0685f187bc/src/libaktualizr/config/base_config.cc#L57

mwasilew replied:
strangely it just worked. I'll check base_config.cc

mike-sul replied:
```
cat z-local.toml | grep -A 4 p11
[p11]
label = "foobar"
[mike@t14s device]$ $AKLITE_EXE -c . staatus --loglevel 0 | grep p11 -A 8
[p11]
module = ""
pass = ""
uptane_key_id = ""
tls_ca_id = ""
tls_pkey_id = ""
tls_clientcert_id = ""

[tls]

```

mike-sul replied:
Afer I added `CopyFromConfig(label, "label", pt);` I get

```
[mike@t14s device]$ $AKLITE_EXE -c . staatus --loglevel 0 | grep p11 -A 8
[p11]
module = ""
pass = ""
label = "foobar"
uptane_key_id = ""
tls_ca_id = ""
tls_pkey_id = ""
tls_clientcert_id = ""
```

doanac reviewed (commented):
nice. once mike is happy, i'm happy

mike-sul commented:
LGTM.

I suppose the PR to `aktualizr-lite` repo should be updated too.

mike-sul commented:
@ricardosalveti 
I suggest to add generation of the .toml snippet to in meta-lmp or even in meta-updater. Something like:

```
/usr/lib/sota/conf.d/46-pkcs-label.toml that defines
[p11]
label = "aktualizr"
```

mike-sul commented:
@pattivacek FYI

ricardosalveti commented:
> @ricardosalveti I suggest to add generation of the .toml snippet to in meta-lmp or even in meta-updater. Something like:
> 
> ```
> /usr/lib/sota/conf.d/46-pkcs-label.toml that defines
> [p11]
> label = "aktualizr"
> ```

I think it is a fair assumption.

@mwasilew want to propose that change as well? We can use the aktualizr recipe (appends) and install when hsm is defined in PACKAGECONFIG (see aktualizr from meta-updater).

mwasilew commented:
in meta-lmp or meta-updater?

mike-sul commented:
Also, I think it makes sense to update lmp-dev-reg in this spot https://github.com/foundriesio/lmp-device-register/blob/959126ae2350851eee33b140b17eb056e77046ac/src/main.cpp#L395 so it adds `label = $hsm_token_label` to the `[p11]` of sota.toml.

Maybe the `hsm_token_label` value should be unique per factory or per device?

ricardosalveti commented:
> in meta-lmp or meta-updater?

for now just in meta-lmp (via our appends), once the ak stuff is merged upstream we can change meta-updater.

ricardosalveti commented:
> Also, I think it makes sense to update lmp-dev-reg in this spot https://github.com/foundriesio/lmp-device-register/blob/959126ae2350851eee33b140b17eb056e77046ac/src/main.cpp#L395 so it adds `label = $hsm_token_label` to the `[p11]` of sota.toml.

+1
 
> Maybe the `hsm_token_label` value should be unique per factory or per device?

I would prefer to default to aktualizr for now, as this will be mainly used by aktualizr/fioconfig at this stage.
